### ISSUE TITLE
`alwayslink`ing a rust_library in a cc_binary

### ISSUE THREAD

**Issue by FaBrand** [bazelbuild/rules_rust#3722]
# Issue
When building a rust_library that contains unreferenced symbols that is a link dependency of a `cc_binary`

e.g. lowlevel CPU startup code that get's located into a defined address via the section.

To ensure that the symbols get linked the `alwayslink` attribut of the library gets set to `True`

<details>
<Summary>Code Example</Summary>

```Starlark
load("@rules_rust//rust:defs.bzl", "rust_library")

rust_library(
    name = "startup_code",
    srcs = glob(["rust/src/**/*.rs"]),
    deps = ["@crates//:vcell"],
    alwayslink = True,
)


cc_binary(
    name = "main",
    deps = [
      ":startup_code",
    ],
)
```

```
# rust/src/startup.rs
#[no_mangle]
#[link_section = ".cpu_entrypoint"]
pub fn _start() -> ! {
    // Startup code
}

#[link_section = ".cpu_entrypoint"]
#[no_mangle]
pub static __ENTRY_POINT: fn() -> ! = _start;
```
</details>

# The error

```
ld.lld failed: error executing CppLink command (from target //:main)
ld.lld: error: bazel-out/platform-dbg/bin/libstartup_code-2667693308.a(lib.rmeta): not an ELF file
```
I assume that the archiver puts the rmeta file into the archive as well.

There are also warnings `<other-archive>.a: archive member 'lib.rmeta' is neither ET_REL nor LLVM bitcode` that hint at this

However with alwayslink we enforce to include the complete archive and the linker failes to include the rmeta file.

I tried my way through the code base. I saw the tests for `alwayslink` but only in combination with rust_binary.

**Comment by FaBrand** [comment]
I found a similar issue in the llvm-project from wasm:
https://github.com/llvm/llvm-project/issues/55786

Here also the `--whole-archive` is passed, leading to the same error

**Comment by krasimirgg** [comment]
I'll investigate a bit in a week or two. Could you share details about your host and target platform and c++ toolchain used? It's possible that some of the linkers `ld.lld` do not support or get tripped on the lib.rmeta.

Internally, we have a custom patch to rules_rust that adds a step that removes the .rmeta file from the generated .rlib archive and presents it as an .a archive to the linker (older versions of bazel or linkers didn't accept .rlib-s as inputs). I'll see what I can do to upstream it.
